Entities/MovableObject.cpp

```diff
@@ -698,37 +697,15 @@ float MovableObject::GetAltitude(int max, int accuracy)
     return g_SceneMan.FindAltitude(m_Pos, max, accuracy);
 }
 
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 
-//////////////////////////////////////////////////////////////////////////////////////////
-// Virtual method:  RestDetection
-//////////////////////////////////////////////////////////////////////////////////////////
-// Description:     Does the calculations necessary to detect whether this MO appears to
-//                  have has settled in the world and is at rest or not. IsAtRest()
-//                  retreves the answer.
-
-void MovableObject::RestDetection()
-{
-    if (m_PinStrength)
-        return;
-
-    // Translational settling detection
-    if ((m_Vel.Dot(m_PrevVel) < 0)) {
-        if (m_VelOscillations >= 2 && m_RestThreshold >= 0)
-            m_ToSettle = true;
-        else
-            ++m_VelOscillations;
-    }
-    else
-        m_VelOscillations = 0;
-
-//    if (fabs(m_Vel.m_X) >= 0.25 || fabs(m_Vel.m_Y) >= 0.25)
-//        m_RestTimer.Reset();
-
-    if (fabs(m_Pos.m_X - m_PrevPos.m_X) >= 1.0f || fabs(m_Pos.m_Y - m_PrevPos.m_Y) >= 1.0f)
-        m_RestTimer.Reset();
+void MovableObject::RestDetection() {
+	if (m_PinStrength) {
+		return;
+	}
+	if (std::abs(m_Pos.m_X - m_PrevPos.m_X) > 1.0F || std::abs(m_Pos.m_Y - m_PrevPos.m_Y) > 1.0F) { m_RestTimer.Reset(); }
```

Comment:
Please delete this axis-dependent physically inconsistent check. Just check against (m_Pos - m_PrevPros).MagnitudeIsGreaterThan()

Also - I have some concerns about this simpler method removing the oscillation check. The old code seems specifically formulated to stop pixels jittering about back-and-forth. Is there a reason this is no longer needed?

Reply:
Honestly your guess is as good as mine about the oscillation check's usage, but in the end it's redundant in comparison to the simpler method, and probably harmful considering that it chooses to hard-settle objects instead of dealing with them through the resting timer and all

Reply:
I'm not sure I agree. I'm pretty sure that the oscillation check is to ensure that objects which are generally stationary but shifting back-and-forth repeatedly by 1 pixel end up settling. With how funky CC's physics is, I'm pretty sure that can happen.

Reply:
I don't buy it. Not only is the code far more complicated than the naive way of checking velocity, the argument that it was originally just for checking if an object is moving or not doesn't really stand up when it's kept around.

This code fundamentally doesn't work if the rest timer stuff is removed. Most objects (that aren't jittering) would never rest in that case, no matter how long they've been still for, because this check fundamentally requires that the velocity has been moving back and forth every single frame for it to trigger. So there's no way it was an original implementation that was replaced with the rest timer.

Furthermore, MOSRotating does the same check but for angular velocity. Which leads me to believe it's very intentional.

Incorporating the rest timer wouln't make much sense, because the constant jittering would reset the rest timer. It's a bit meaningingless in that instance. And, again to mention - the Lua AI was previously broken because of a not dissimilar change to how the stuck calculation was performed.

As such, I'm not particularly comfortable with this change.

Reply:
As more evidence the rest detection was added later - the RestTimer is near the top of the private members and it's one of the very few things that is uncommented. The Oscillations is near the bottom of the private members (in both Vel and Ang) is near the button, and it commented. 

That to me implies that the oscillations check was added *after* the rest timer because it wasn't enough.

Reply:
Alright, as for incorporating the rest timer, what do you mean? Couldn't the oscillation check be used to just bypass the reset?

Reply:
I don't understand, why can't it be integrated into the same condition as the one that resets the timer?

Reply:
Because, it's not about the dot product. It's not about changing direction just once. This is keeping track of if the particle repeatedly changed direction, and if so, forces the particle to rest.

Your code would do the opposite - it would require that a particle *must* be swapping direction to rest. It's not guarding against an exceptional case as the original does, instead your code there is making that jittering behaviour a fundamental requirement for the rest to occur. Furthermore it's requiring that this jittering must be occuring for at least the length of the rest timer.

Reply:
Whoops, my bad! 

Do you think it's detrimental that two oscillations is enough to trigger this? Wouldn't three be a better indicator of something being truly wonky?

Reply:
I'm not sure. Overall I'm not too inclined to change it unless the consequences are properly tested. I'd imagine that whoever wrote this code tested 2 and thought it was reasonable.